### DISCUSSION TITLE
Feature request: Way to add hetzner dedicated server instance as agent node

### DISCUSSION THREAD

**Discussion by yashodhank** [mysticaltech/terraform-hcloud-kube-hetzner#283] [Ideas]
It would be awesome if you consider ability to automatically deploy vswitch with dedicated server to hetzner cloud instances (join the cluster).

I am still learning and I really love the way you all have built this amazing project.
I have experimented with it and so far and it worked effortlessly. Thank you so much for that..
I am looking forward to contribute too.

**Comment by phaer** [comment]
As far as I am aware, there's no terraform provider for neither vswitches nor dedicated servers we could use for that feature. At least https://registry.terraform.io/providers/hetznercloud/hcloud/ does not support them. And even if there was, I don't think you can use vswitches with hcloud hosts?
 As long as that's the case I am afraid this feature request is out of scope.

**Comment by yashodhank** [comment]
I agree, there is no official support yet as it is with hcloud.

but there are few working implementations 
https://registry.terraform.io/providers/Peters-IT/hetzner-robot/latest
https://github.com/Peters-IT/terraform-provider-hetzner-robot

**Comment by phaer** [comment]
Did you test it? From the docs it looks like it neither supports provisioning servers nor attaching hcloud_networks to vswitches (just reading them). If it works, I think a first step would be to implement a shared network on top of kube-hetzner, the only change to kube-hetzner required would be to expose the network_ids, to attach them to the vswitch, right? 

Joining dedicated machines into your cluster would probably left as an exercise to the reader as it depends on your provision those servers. That would be an interesting project in itself, but probably best implemented on top of kube-hetzner (re-using its terraform modules for the hcloud part).

Let us know if you start hacking on this :)

**Comment by yashodhank** [comment]
Sure, valuable feedback. :)
I have seen couple of tutorials explaining how to connect vswitch vlan and hcloud for private network connection between root servers and hcloud instances.

I am going to test it in next couple of days. 

Ref Docs:
https://robot.your-server.de/doc/webservice/en.html#preface

Interesting project yesterday came across (it implements HCLOUD + ROOTServer API):
https://github.com/syself/cluster-api-provider-hetzner/blob/main/docs/topics/preparation.md

**Comment by phaer** [comment]
Very interesting, thanks! I am renting a bunch of bigger dedicated machines and two vswitches from Hetzner myself (albeit not running kubernetes there) and learning that you can, at least nowadays, join them into private networks with hcloud machines is quite interesting to me.

> https://github.com/syself/cluster-api-provider-hetzner/blob/main/docs/topics/preparation.md

That one looks very interesting indeed, hadn't heard of it so thanks a lot!

**Comment by mysticaltech** [comment]
Moving it to a discussion where it makes more sense, in the ideas section.

**Comment by GerardGarcia** [comment]
if I understand correctly, kube-hetzner doesn't support having hybrid cluster with cloud and dedicated machines but that doesn't mean that it is not doable given that it is possible to join cloud and dedicated machines in the same private network. 

I've just started to look into setting up a k3s cluster using kube-hetzner but in the future I'd like to have the possibility to add dedicated servers to the mix, even if I have to do it manually or automate it myself. Since I've just started I'm not sure how hard it would be, so if I may ask: do you foresee any pain points in particular if I ever want to add a dedicated server to the cluster and configure it myself? which I assume will involve add a new node to the k3s cluster, configure the load balancer (if necessary, not sure if being it the same private network may be enough), configure longhorn, among other things.

**Reply by mysticaltech** [reply]
@GerardGarcia I cannot comment too much as I have not done it myself. But if you search this project, some users have pulled this off. So it's doable. However, not something we can support today, as too much custom work, would be hard to maintain.

**Reply by GerardGarcia** [reply]
no problem, I'll do some research and tests. Knowing that is doable hopefully I'll find a way :)

**Comment by Aubermean** [comment]
A discussion has been opened here: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/1311

**Comment by mysticaltech** [comment]
Now supported see, https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/master/docs/add-robot-server.md
